### DISCUSSION TITLE
Best Practices for SkyWalking Architecture for Multi-Cluster Kubernetes

### DISCUSSION THREAD

**Discussion by cahyahikmawan** [apache/skywalking#12588] [Q&A] [answered]
Hello, I'm new to using SkyWalking, and I want to make sure I choose the best option for implementing it in my production environment.

For reference, my clusters are:
- infra-tools-cluster
- app-group1-cluster
- app-group2-cluster

I have deployed SkyWalking OAP and SkyWalking UI in the `infra-tools-cluster`. Everything works perfectly when I deploy a demo app inside `infra-tools-cluster`. I tested cross-cluster communication with SkyWalking OAP using two methods, both of which worked properly:

1. Assigning a `kind: Service` to SkyWalking OAP with a `LoadBalancer` type.
2. Using ingress-nginx with the `--tcp-services-configmap` flag enabled in the ingress-nginx deployment to proxied port 11800 to Skywalking OAP Service.

Are these good options for running in production? Do you have any recommendations for managing **multi-cluster** communication? Is SkyWalking Satellite designed for this purpose?

SkyWalking Version: 10.0.1
Elasticsearch Version: 8.14.3
Deployment Mode: Kubernetes (GKE)

Thank you.

**Comment by wu-sheng** [comment]
What scenarion are you using? Agent, service mesh or something?

**Reply by cahyahikmawan** [reply]
Thanks for the quick response.

I plan to use the `agent` to send data to SkyWalking OAP. 
Quick question: with a service mesh, do I still need to inject an agent like in Istio?

**Reply by wu-sheng** [reply, marked as answer]
No, service mesh obs doesn't rely on agent at all.

And how to do load balancer cross clusters is your choice. We don't recommend foe that.

Satellite is also for in cluster balancing.

The important thing is, the agent(Java) is using grpc, which is a keep-connection http2. So, the most load balancer is connection oriented,not data oriented.

**Reply by cahyahikmawan** [reply]
Okay, I got it. I appreciate your help. 👍 Thanks!

**Comment by wokwong** [comment]
Hi @cahyahikmawan , what kind of architecture did you finally choose ？

**Reply by cahyahikmawan** [reply]
hi @wokwong  Since I need to pay for every type of LoadBalancer in GCP, I decided to use an existing middleware (Ingress-NGINX). However, be aware that using annotations in kind: Ingress does not work well with gRPC, which is why I’m using --tcp-services-configmap. I’m not completely sure, but it seems similar to how an established connection works with Kafka.

**Reply by wokwong** [reply]
So, you use the same SkyWalking for multiple environments,right?

**Reply by wokwong** [reply]
I'm considering whether to use a single SkyWalking for all environments, or to deploy a separate SkyWalking system for each environment.
